### PR TITLE
feat: support hot_restart

### PR DESCRIPTION
**Pull request by bobozhengsir**
## Motivation

Hot restart (also named `smooth process upgrade`) refers to the ability of a system or software to undergo upgrades or updates without disrupting the ongoing processes or operations. Supporting smooth process upgrade is essential for maintaining system availability, minimizing downtime, and providing a seamless user experience during the upgrade process.

## Solution

The key to hot restart is how to enable two different processes to share the same TCP or UNIX-domain socket, especially when they are not parent-child processes. Fortunately, Linux (or rather UNIX) provides us with the necessary tool. We can use UNIX-domain sockets to pass file descriptors between applications, specifically utilizing the SCM_RIGHTS mechanism.

Our solution focuses on the following key points:

1. To utilize hot restart, you need to initialize the hot restart mechanism.
2. The process first attempts to connect to the parent_sock. If the connection fails, it means that the process is the original parent process. In this case, the process should bind the parent_sock and parent_handle, waiting for the child process to send a message.
3. If the child process successfully connects to the parent_sock, it duplicates the file descriptors using `dup_parent_listener_sock`.
4. Once all listener sockets are duplicated, the child process sends a terminate_parent request to the parent process.
5. The parent process receives the termination signal (term) and gracefully exits."

### CODE REVIEW

File: volo/src/hotrestart/mod.rs [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+const HOT_RESTART_CHILD_ADDR: &'static str = "volo_hot_restart_child.sock";
+
+lazy_static! {
+    pub static ref DEFAULT_HOT_RESTART: HotRestart = HotRestart::new();
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Debug)]
+#[repr(u8)]
+enum HotRestartState {
+    Uninitalize = 0,
```
Comment (PureWhiteWu) [inline volo/src/hotrestart/mod.rs:42]: Uninitalized

File: volo/src/hotrestart/mod.rs [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+    ParentInitialized,
```
Comment (PureWhiteWu) [inline volo/src/hotrestart/mod.rs:43]: 后面的序号明确标出来？

File: volo/src/hotrestart/mod.rs [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+    os::fd::{AsRawFd, RawFd},
+    path::{Path, PathBuf},
+    sync::{
+        atomic::{AtomicI32, Ordering},
+        Arc, Mutex as StdMutex, OnceLock,
+    },
+    time::Duration,
+};
+
+use futures::lock::Mutex;
```
Comment (PureWhiteWu) [inline volo/src/hotrestart/mod.rs:15]: 这里为什么需要用到这个 Mutex？

File: volo/src/hotrestart/mod.rs [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+
+// simple self message
+enum HotRestartMessage {
+    TerminateParentRequest,
+    PassFdRequest(String),
+    PassFdResponse(RawFd),
+}
+
+pub struct HotRestart {
+    state: Arc<Mutex<HotRestartState>>,
```
Comment (PureWhiteWu) [inline volo/src/hotrestart/mod.rs:97]: 这里的 HotRestartState 是 u8，感觉可以改成用 AtomicU8？

File: volo/src/hotrestart/mod.rs [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+    sync::{
+        atomic::{AtomicI32, Ordering},
+        Arc, Mutex as StdMutex, OnceLock,
+    },
+    time::Duration,
+};
+
+use futures::lock::Mutex;
+use lazy_static::lazy_static;
+use nix::{
```
[the remark this answers is not in the document]
Comment (bobozhengsir) [reply volo/src/hotrestart/mod.rs:17]: 整个包是unix才使用的，lib.rs cfg注明了

File: volo/src/hotrestart/mod.rs (hunk 1 above)
Comment (bobozhengsir) [reply volo/src/hotrestart/mod.rs]: done

File: volo/src/hotrestart/mod.rs (hunk 2 above)
Comment (bobozhengsir) [reply volo/src/hotrestart/mod.rs]: done

File: volo/src/hotrestart/mod.rs (hunk 3 above)
Comment (bobozhengsir) [reply volo/src/hotrestart/mod.rs]: 换成tokio::sync::Mutex了，domain sock io操作跨await使用

File: volo/src/hotrestart/mod.rs (hunk 4 above)
Comment (bobozhengsir) [reply volo/src/hotrestart/mod.rs:97]: state有互斥性要求，不是计数使用，用Mutex保护下

**Review by PureWhiteWu — APPROVED** [review]

**Review by Millione — APPROVED** [review]

**Pull request merged**